extensions/binding/org.eclipse.smarthome.binding.lirc/src/main/java/org/eclipse/smarthome/binding/lirc/internal/LIRCHandlerFactory.java

```diff
@@ -0,0 +1,89 @@
+
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.binding.lirc.LIRCBindingConstants;
+import org.eclipse.smarthome.binding.lirc.handler.LIRCBridgeHandler;
+import org.eclipse.smarthome.binding.lirc.handler.LIRCRemoteHandler;
+import org.eclipse.smarthome.binding.lirc.internal.discovery.LIRCRemoteDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link LIRCHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Andrew Nagle - Initial contribution
+ */
+public class LIRCHandlerFactory extends BaseThingHandlerFactory {
+
+    @SuppressWarnings("unused")
+    private Logger logger = LoggerFactory.getLogger(LIRCHandlerFactory.class);
+
+    private final static Set<ThingTypeUID> SUPPORTED_THING_TYPE_UIDS = Sets.union(
+            LIRCBindingConstants.SUPPORTED_DEVICE_TYPES,
+            LIRCBindingConstants.SUPPORTED_BRIDGE_TYPES);
+
+    private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPE_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (LIRCBindingConstants.SUPPORTED_BRIDGE_TYPES.contains(thingTypeUID)) {
+            LIRCBridgeHandler handler = new LIRCBridgeHandler((Bridge) thing);
+            registerDeviceDiscoveryService(handler);
+            return handler;
+        } else if (thingTypeUID.equals(THING_TYPE_REMOTE)) {
+            return new LIRCRemoteHandler(thing);
+        }
+
+        return null;
+    }
+
+    @Override
+    protected void removeHandler(ThingHandler thingHandler) {
+        if (this.discoveryServiceRegs != null) {
+            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(thingHandler.getThing().getUID());
+            if (serviceReg != null) {
+                serviceReg.unregister();
+                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
+            }
+        }
+    }
+
+    private void registerDeviceDiscoveryService(LIRCBridgeHandler handler) {
+        LIRCRemoteDiscoveryService discoveryService = new LIRCRemoteDiscoveryService(handler);
+        discoveryService.activate();
```

Comment:
I don't have a whole lot of experience with OSGi, what would the proper way of doing this be?